clang/tools/driver/driver.cpp

```diff
@@ -331,6 +340,56 @@ static void SetInstallDir(SmallVectorImpl<const char *> &argv,
   // path being a symlink.
   SmallString<128> InstalledPath(argv[0]);
 
+#if defined(__linux__)
```

Comment:
Can we use [`getMainExecutable()`](https://github.com/llvm/llvm-project/blob/1f0126902876cb42a7502047e62bdf512fd96b02/llvm/include/llvm/Support/FileSystem.h#L1345) instead?

(This sort of system-specific logic should live in LLVM's Support directory so that we don't sprinkle host-specific preprocessor conditionals all over, but better still if we can use our existing facilities rather than invent new ones.)

Reply:
I was kind of thinking the same. And then started to wonder about why SetInstallDir says "We do this manually, because we want to support that path being a symlink.". I don't really understand if that indicates that the installed dir should be set based on the symlinks path or not.

Afaict `Driver::ClangExecutable` already is computed via `getMainExecutable()`, right. 
And the Driver c'tor is setting both `Driver::Dir` and `Driver::InstalledDir` based on the path of the clang executable.

So is the problem here is that SetInstallDir is called after the above (having contructed `TheDriver`). Resulting in `InstalledDir` being changed into being relative to the symlink?

Reply:
> Can we use `getMainExecutable()` instead?

Good point! 

> So is the problem here that SetInstallDir is called after the above (having constructed TheDriver). Resulting in InstalledDir being changed into being relative to the symlink?

When I first encountered this problem I tried to run a debug session to understand how things work, but got confused and in the end I patched `SetInstallDir` and for my needs it was ok, but since `getMainExecutable()` already does the job I'll reconsider and try to understand where the problem comes from.